File: 002-greatest-common-divisors/Solution.py
class Solution:
    # https://leetcode.com/problems/greatest-common-divisor-of-strings/submissions/
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if len(str1) == len(str2):
            return str1 if str1 == str2 else ""

        if len(str1) > len(str2):
            if not str1.startswith(str2):
                return ""
            return self.gcdOfStrings(str1[len(str2):], str2)
        
        if len(str2) > len(str1):
            if not str2.startswith(str1):
                return ""
            return self.gcdOfStrings(str2[len(str1):], str1)

        return ""

File: 002-greatest-common-divisors/Solution.java
class Solution {
    //https://leetcode.com/problems/greatest-common-divisor-of-strings/submissions/
    public String gcdOfStrings(String str1, String str2) {
        if(str1.equals(str2)) {
            return str1;
        }
        if(str1.length() > str2.length()) {
            if(!str1.startsWith(str2)) {
                return "";
            }
            return gcdOfStrings(str1.substring(str2.length()), str2);
        }

        if(str2.length() > str1.length()) {
            if(!str2.startsWith(str1)) {
                return "";
            }
            return gcdOfStrings(str2.substring(str1.length()), str1);
        }
        return "";
    }
}